### PR TITLE
YJIT (arm64): fix off-by-one in the single-branch range check

### PR DESCRIPTION
**Pull request by felixbuenemann**
`jmp_ptr_bytes()` reserves room for a patchable jump -- it sets page_end_reserve and bounds conditional jump width. A single `b` suffices when every branch in the code region is in imm26 range; otherwise it reserves 5 instructions for an absolute load-address plus br.

The range test was `virtual_region_size() / 4`, but two instructions in an S-byte region are at most S-4 bytes apart. At exactly 128MiB, S/4 is 2^25 while the widest imm26 offset is 2^25-1, so the check gave up one instruction early and took the 5-instruction fallback -- despite the comment right above it saying <= 128 should work.

That boundary is the default. exec_mem_size falls back to mem_size in `CodegenGlobals::init` and mem_size defaults to 128MiB, so every arm64 run without an explicit `--yjit-exec-mem-size` reserved 20 bytes per patchable jump where 4 would do. On a branch-heavy method that is the difference between 1336 and 956 bytes of inline code.

Test (S-4)/4 instead, and add a unit test pinning both sides of the boundary. `--yjit-exec-mem-size=129` still takes the fallback, as before.

Checked with `make yjit-check` on arm64: 374 cargo tests, all 2048 bootstraptest at `--yjit-call-threshold=1`, and test_yjit.rb (142 tests, 0 failures). Since the default size is the boundary case, the whole suite ran on the tightened reservation.

This bugfix was prepared with the help of AI coding tools, among them a copilot comment on the same bug for ZJIT in PR #17992.

### CODE REVIEW

**Comment by XrXr** [comment]
This is just wrong. Think about a region that's 4 bytes, which fits the `B`. The `B` can reach every instruction in the region.

**Comment by felixbuenemann** [comment]
@XrXr Please explain why you think it is wrong?

**Comment by felixbuenemann** [comment]
@XrXr Also see my comment https://github.com/ruby/ruby/pull/17992#discussion_r3624596752 I don't think this PR is wrong.

**Comment by felixbuenemann** [comment]
But I have a test fix I need to push here to avoid large memory usage in tests, which causes the previous yjit test failures.

**Comment by felixbuenemann** [comment]
The reason the PR matters is because the bug happens at exactly 128 MiB, which is the default YJIT memory size used by Ruby. So it slows down YJIT for most arm64 users.

**Comment by XrXr** [comment]
🤦🏼 Apologies, I clicked send too quick. I was wrong. The piece that made it click for me was that the address of the last instruction in the region is `region_size/4 - 1`, and the largest offset is the same from address 0. Testing the positive side makes sense because that handles the two's complement minimum.

Can you force push to what's on this PR so we can reopen it? Sorry again for the hassle.

 
Screenshot 2026-07-21 at 15 05 10

**Comment by felixbuenemann** [comment]
@XrXr I do not have that option. Should I resubmit as a new PR and link it here?

**Comment by felixbuenemann** [comment]
@XrXr or do you mean force push the commit from this PR so *you* can reopen it?

**Comment by felixbuenemann** [comment]
I restored the `fix-yjit-arm64-branch-range-off-by-one` to commit `1c8a0bbdc14bdda7bb06efd9b666d65b9a8eb8aa`, I think that is what you wanted me to do.

I will push the test fix, once it it reopened.

File: yjit/src/virtualmem.rs [hunk 1]
```diff
@@ -117,13 +118,23 @@ pub enum WriteError {
...
 
 impl<A: Allocator> VirtualMemory<A> {
+    /// Whether mapping `required_region_bytes` keeps us inside the memory limit.
+    /// A None limit is unbounded, so tests can map pages in an oversized region
+    /// without the process-wide yjit_alloc_size() getting in the way.
+    fn within_memory_limit(&self, required_region_bytes: usize) -> bool {
+        match self.memory_limit_bytes {
+            Some(limit) => required_region_bytes + yjit_alloc_size() < limit,
+            None => true,
+        }
```
[the remark this answers is not in the document]
Comment (felixbuenemann) [reply yjit/src/virtualmem.rs]: The code is preexisting, it was moved from `(page_addr + page_size - start as usize) + yjit_alloc_size() < self.memory_limit_bytes` into `within_memory_limit()` and if we really execed 18 exabytes (2^64), the code would paning in both debug and release builds due to `-C overflow-checks=on`.

File: yjit/src/backend/arm64/mod.rs [hunk 2]
```diff
@@ -1382,6 +1383,20 @@ mod tests {
...
+        // still the largest offset a b can encode, so one instruction is enough.
+        // This is the default region size, so it is the common case.
+        let cb = CodeBlock::new_dummy(128 * 1024 * 1024);
+        assert_eq!(4, cb.jmp_ptr_bytes());
+
+        // One page further and a b can no longer span the region, so we have to
+        // reserve room for an absolute load-address plus br.
+        let cb = CodeBlock::new_dummy(128 * 1024 * 1024 + 16 * 1024);
+        assert_eq!(20, cb.jmp_ptr_bytes());
+    }
```
[the remark this answers is not in the document]
Comment (felixbuenemann) [reply yjit/src/backend/arm64/mod.rs:1402]: `vec![0; mem_size]` goes through alloc_zeroed -> calloc -> anonymous mmap, which hands back lazily-zeroed pages. The test only reads `virtual_region_size()` and never writes, so none of those pages is ever faulted in.

File: yjit/src/backend/arm64/mod.rs [hunk 3]
```diff
@@ -1382,6 +1383,20 @@ mod tests {
...
+        // The widest branch in a 128MiB region spans 128MiB - 4 bytes, which is
+        // still the largest offset a b can encode, so one instruction is enough.
+        // This is the default region size, so it is the common case.
+        let cb = CodeBlock::new_dummy(128 * 1024 * 1024);
+        assert_eq!(4, cb.jmp_ptr_bytes());
+
+        // One page further and a b can no longer span the region, so we have to
+        // reserve room for an absolute load-address plus br.
+        let cb = CodeBlock::new_dummy(128 * 1024 * 1024 + 16 * 1024);
+        assert_eq!(20, cb.jmp_ptr_bytes());
```
[the remark this answers is not in the document]
Comment (felixbuenemann) [reply yjit/src/backend/arm64/mod.rs]: Amended the proposed fix.

File: yjit/src/virtualmem.rs [hunk 4]
```diff
@@ -33,7 +33,8 @@ pub struct VirtualMemory<A: Allocator> {
     region_size_bytes: usize,
 
     /// mapped_region_bytes + yjit_alloc_size may not increase beyond this limit.
-    memory_limit_bytes: usize,
+    /// None disables the limit, which tests use to build oversized regions.
+    memory_limit_bytes: Option<usize>,
```
Comment (XrXr) [inline yjit/src/virtualmem.rs]: Let's not introduce optionality here solely for test. Let's drop this commit and find a different solution to fix the test.

**Review by XrXr — CHANGES_REQUESTED** [review]

File: yjit/src/virtualmem.rs (hunk 4 above)
Comment (felixbuenemann) [reply yjit/src/virtualmem.rs]: Solved by dropping that commit and splitting ` jmp_ptr_bytes_for_region()` into a private and public method, so we can test the private version without allocating memory. Also rebased on latest master.

**Comment by felixbuenemann** [comment]
@XrXr Anything else left to do here? I think I addressed all your feedback.

**Review by XrXr — APPROVED** [review]
